### PR TITLE
Switch to latest vm-memory and tweak the notification logic

### PR DESCRIPTION
**Pull request by alexandruag**
This PR switches to using version `0.3.0` of `vm-memory` which adds atomic `load` and `store` support as part of the `Bytes`/`GuestMemory` interface. There are also some proposed changes around the notification logic (for example, double checking when enabling notifications to make sure we don't miss one), as well as more comments just in case they are helpful to people reading the code (there's one example around enabling notifications that should be turned into a doc comment/example for consumers of the crate, because it's not  immediately clear what the use pattern looks like).

Also, I was wondering if it makes sense to remove `DescriptorChainRwIter` and the related logic, because it looks like we can achieve the same by calling `filter()` on `DescriptorChain` when iterating. WDYT?

### CODE REVIEW

**Comment by slp** [comment]
I've just taken a first look and I like it. I'll look at it in depth tomorrow, and will also play a bit with a couple of vhost-user devices consuming this crate.

**Review by bonzini — DISMISSED** [review]
Looks good to me, though I have not done a careful review beyond the orderings and fences.

I like having Ordering arguments even though they should always be Relaxed , because it tells you "this is using atomic accesses, look for fences" (and likewise for Release "do not look for fences". I had never seen this idiom but it's very self-documenting!

File: src/queue.rs
Comment (andreeaflorescu) [inline src/queue.rs:248]: Why is this `unchecked_add` safe?

File: src/queue.rs
Comment (andreeaflorescu) [inline src/queue.rs:316]: Mind to also say why it cannot overflow? :)) Otherwise, we can just remove the comment.

File: src/queue.rs
Comment (andreeaflorescu) [inline src/queue.rs:320]: Hmm, this assumption has the potential to introduce time of check, time of use bugs. Are we making sure in anyway that next is only called after `queue::is_valid`?

**Review by andreeaflorescu — COMMENTED** [review]
Just some questions. Didn't fully review it.

**Review by slp — DISMISSED** [review]
Took a second look and still LGTM (apart from the issues pointed by @andreeaflorescu). I've tested the notification logic with `virtiofsd-rs` and interrupt suppression works just as expected. The API changes are reasonable. We should probably wait for @jiangliu 's feedback before merging this, though.

On the other hand, we should probably think about a versioning policy for API changes.

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:248]: Added a comment.

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:316]: Added a bit more details to the comment.

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:320]: We do not currently enforce that a queue is only used after checking `is_valid`, but rather expect the device implementations to do that before activation. The standard also forbids drivers to change queue parameters while the device is "running". A warp-around cannot lead to unsafe memory accesses because the memory model will check everything. We can add more validations if that seems like a worthwhile improvement, but I'm not sure if the best place is here or as part of logic that mutates the queue configuration.

**Review by alexandruag — COMMENTED** [review]
Thanks for the reviews! Having a policy around API (breaking) changes is definitely something we should have in mind, but I think many components will still go through some churn until reaching an initial stable state. Any ideas/guidelines to use as a starting point for that discussion would be great though.

@jiangliu, can you please have a look as well?

File: src/queue.rs
Comment (andreeaflorescu) [reply src/queue.rs:325]: The explanation from your comment should also be found in the code IMHO.

**Comment by slp** [comment]
@jiangliu I'd like to merge this one to be able to update `vhost-user-backend` to use `vm-memory` 3.x. Could you please take a look?

File: src/queue.rs
Comment (jiangliu) [inline src/queue.rs:686]: Why the "- Wrapping(1u16)" has been removed?

**Comment by jiangliu** [comment]
> @jiangliu I'd like to merge this one to be able to update `vhost-user-backend` to use `vm-memory` 3.x. Could you please take a look?

Sorry for reply, it took some time to figure out the indirect related question:(
It looks good to me except the "- Wrapping(1u16)" doubt mentioned above.

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:686]: It hasn't been removed entirely, it's now part of the computation at the beginning of the function (line 672) which computes the last index in the used ring where we wrote something.

Comment (jiangliu) [reply src/queue.rs:686]: Seems it still has changed the logic because the `prev` is self.next_used - Wrapping(1) of last round?

File: src/queue.rs
Comment (jiangliu) [inline src/queue.rs:1330]: The spec states:
_For example, if used_event is 0, then a device using VIRTIO_F_EVENT_IDX would send a used buffer notification to the driver after the first buffer is used (and again after the 65536th buffer, etc)._

So how about adding a case for "again after the 65536th buffer"?

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:686]: Oh, I see your point now. So essentially, the left-hand side of the inequality is the same as before (because `used_event` is the same, and `used_idx` is now defined as `self.next_used - Wrapping(1)`), but the right-hand side is different because `used_idx` is now the old `used_idx` with a `- Wrapping(1)` as well. After a bit of doodling I think you are right and the change is erroneous. We should use a `>` instead of `>=` to account for the implicit `- Wrapping(1)` so things are exactly as before.

This also has the added advantage that `needs_notification` will return true until `used_idx` actually changes, and I think we can use a value that defaults to 0 for `signalled_used` instead of an `Option`. I'll experiment a bit more with the changes and update the PR later today or tomorrow at the latest. Thanks for pointing this out!

File: src/queue.rs
Comment (jiangliu) [inline src/queue.rs:1335]: A little hesitating whether it should be "true" or "false" here:)

File: src/queue.rs
Comment (jiangliu) [reply src/queue.rs:686]: Be careful about the ">=" vs ">" and wrapping around issue:)
All implementation of algorithm I have seen using the form "if (used_idx - used_event - Wrapping(1u16)) >= (used_idx - old_idx)".

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:1330]: added

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:686]: Ok so after more playing around, the sad conclusion is there's no nice way of making `needs_notification` idempotent. I think a cleaner solution is something like making `add_used` return a boolean value that indicates whether a notification is needed or not. For the time being, I've changed the code here to be essentially the same as before the  PR, for ease of re-reviewing this part.

Also, turns out the regression caused by the original version of the PR was not around the handling of `- Wrapping(1)` (which actually worked out in the end), but the fact that `signalled_used` was only updated when the method would return `true` (which fit the name, but changed the logic).

File: src/queue.rs
Comment (alexandruag) [reply src/queue.rs:1335]: I think it should be `false` (and it's what the method actually returns now after changing `needs_notification` back).

**Review by alexandruag — COMMENTED** [review]
Hi! I've addressed all outstanding comments. In particular, the `needs_notification` method now effectively implements the same checks as before. We can consider other options after this PR is merged. I'll create an issue about potential future queue refinements, where we can also discuss the aspect of breaking changes to the interface going forward.

**Review by jiangliu — APPROVED** [review]

**Comment by alexandruag** [comment]
@slp can you please have another look? The delta between the last update (following Gerry's comments) and the previous version of the PR is available [here](https://github.com/rust-vmm/vm-virtio/compare/eab0e9b32bbd3899ac53cb4ccf0096c34eebde94..4083ec5488f0aab0fbd0461a92904145f60b6ad5).

**Review by slp — APPROVED** [review]
LGTM.

**Pull request merged**
